Code quality

trueqbit opened the pull request:
New Visual C++ 18 2026 as well as Clang 21 are out. In an attempt to compile sqlite_orm with these new toolchains I corrected and improved things along the way.

Fixes:
* Corrected deprecated spacing in user-defined literal operators, e.g. `operator"" _cte()` -> `operator""_cte()`.

Generic improvements:
* Updated to clang-format 20.
* Used the single-argument version of `static_assert`, which was not possible with C++14 before.
* Removed a few unused type aliases in unit tests.
* Addressed a few warnings about unused variables.

Clang related:
* Temporarily disable spurious Clang warnings instead of suppressing them in code, which makes code cleaner and less likely to forget to suppress them on spot.
* Clang understands C++26 features even in C++17 mode, but it doesn't mean it understands other language features, so we need to be specific using them.
* Needed to disable compiling the R*Tree module with Clang because Clang chokes on using constexpr variables in lambdas. This is an unfixed issue reported several times to the Clang maintainers.

Clang MSVC:
* Made sqlite_orm compile with Clang MSVC. Specifically it still cannot use functions exported from DLLs as compile-time parameters, so we provide workarounds for it.

trueqbit committed:
`stream_identifier()` with an index sequence needs to be available

trueqbit committed:
Used the single-argument version of `static_assert`

trueqbit committed:
Accommodated for Clang on Windows, which cannot use constexpr DLL exports

trueqbit committed:
Updated to clang-format 20

trueqbit committed:
Space in user-defined literal operator is deprecated

trueqbit committed:
Updated and improved code after compiling with Clang MSVC v21 and C++26

* Temporarily disable spurious Clang warnings instead of suppressing them in code.
* Clang understands C++26 even in C++17 mode, but it doesn't mean it understands other language features, so we need to be specific using them.
* Removed a few unused type aliases in UTs.
* Addressed a few warnings about unused variables.
* Needed to disable the rtree module example because Clang chokes on using constexpr variables in lambdas.

trueqbit committed:
Ignore Clang's "-Wunused-lambda-capture" as this is common in generic code

trueqbit committed:
Reordered test expressions for checking Clang MSVC

trueqbit committed:
appveyor: Try using clang-14 for C++20 build on Ubuntu worker image

clang is using libstd++ by default and chokes when including <chrono>.

fnc12 commented on dev/storage.h:
```diff
@@ -1656,7 +1656,7 @@ namespace sqlite_orm {
                 }
 
-                switch (int rc = sqlite3_step(stmt)) {
+                switch (/*int rc =*/sqlite3_step(stmt)) {
```
this thing makes it harder to debug and watch returned value from `sqlite3_step` sometimes

trueqbit replied:
You are right, I think we can just use `[[maybe_unused]]` in such cases.

trueqbit committed:
Marked unused sqlite3 API return codes `[[maybe_unused]]`

trueqbit committed:
Revert "appveyor: Try using clang-14 for C++20 build on Ubuntu worker image"

This reverts commit 7577c9cb506d3b2a8b2b02aecc8d2eb11534bfa6.

trueqbit committed:
appveyor: Used Ubuntu1804 explicitly

The "Ubuntu" tagged worker image is now an alias for Ubuntu2204. We still want to stay at the older toolchain in order to catch buggy compiler behavior.

trueqbit committed:
`[[maybe_unused]]` in a switch init-stmt only supported since msvc 1920

trueqbit committed:
Simplified 'xdestroy' unit test file

trueqbit committed:
Updated previous commit
